net/listener_split.go

```diff
@@ -0,0 +1,110 @@
+package net
+
+import (
+	"context"
+	"net"
+	"net/http"
+
+	"github.com/drand/drand/protobuf/drand"
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/soheilhy/cmux"
+	"google.golang.org/grpc"
+)
+
+// NewGRPCListenerForPublicAndProtocol creates a new listener for the Public and Protocol APIs over GRPC with no TLS.
+func NewGRPCListenerForPublicAndProtocol(addr string, s Service, opts ...grpc.ServerOption) Listener {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		panic("tcp listener: " + err.Error())
+	}
+	mux := cmux.New(l)
+	opts = append(opts, grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor))
+	opts = append(opts, grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor))
+	grpcServer := grpc.NewServer(opts...)
+	g := &grpcListener{
+		Service:    s,
+		grpcServer: grpcServer,
+		mux:        mux,
+		lis:        l,
+	}
+	drand.RegisterProtocolServer(g.grpcServer, g.Service)
+	drand.RegisterPublicServer(g.grpcServer, g.Service)
+	grpc_prometheus.Register(g.grpcServer)
+	return g
+}
+
+type grpcListener struct {
+	Service
+	grpcServer *grpc.Server
+	mux        cmux.CMux
+	lis        net.Listener
+}
+
+func (g *grpcListener) Start() {
+	l := g.mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
+	go g.grpcServer.Serve(l)
+	g.mux.Serve()
+}
+
+func (g *grpcListener) Stop() {
+	g.lis.Close()
+	g.grpcServer.Stop()
+}
+
+// NewRESTListenerForPublic creates a new listener for the Public API over HTTP/JSON without TLS.
+func NewRESTListenerForPublic(addr string, s Service, opts ...grpc.ServerOption) Listener {
```

Comment:
can this one use the http servlet in https://github.com/drand/drand/pull/257 that sets cache headers

Reply:
Out of scope of this PR. Let's do this separately to unblock Hector.